File: sdk-go/tests/bdd/account_steps.go
package bdd

import (
	"fmt"

	"github.com/cucumber/godog"
)

// AccountSteps implements BDD steps for account management scenarios.
type AccountSteps struct {
	testWorld *TestWorld
}

// NewAccountSteps creates a new AccountSteps instance.
func NewAccountSteps(world *TestWorld) *AccountSteps {
	return &AccountSteps{testWorld: world}
}

// RegisterSteps registers all account steps with godog.
func (s *AccountSteps) RegisterSteps(ctx *godog.ScenarioContext) {
	ctx.Given(`^I have an initialized Decibel read client$`, s.givenReadClient)
	ctx.Given(`^I have a subaccount address ([^"]*)$`, s.givenSubaccountAddress)
	ctx.When(`^I request the account overview for a subaccount$`, s.requestAccountOverview)
	ctx.Then(`^I should receive the account overview data$`, s.shouldReceiveAccountOverview)
	ctx.And(`^the overview should include the total margin$`, s.checkTotalMargin)
	ctx.And(`^the overview should include the unrealized PnL$`, s.checkUnrealizedPnl)
	ctx.And(`^the overview should include the cross margin ratio$`, s.checkCrossMarginRatio)
	ctx.When(`^I request the account positions$`, s.requestPositions)
	ctx.Then(`^I should receive the account positions$`, s.shouldReceivePositions)
	ctx.When(`^I request the account open orders$`, s.requestOpenOrders)
	ctx.Then(`^I should receive the open orders$`, s.shouldReceiveOpenOrders)
	// Additional steps can be added following the same pattern
}

func (s *AccountSteps) givenReadClient() error {
	_, err := s.testWorld.GetReadClient()
	if err != nil {
		s.testWorld.SetError(err)
		return err
	}
	return nil
}

func (s *AccountSteps) givenSubaccountAddress(addr string) error {
	s.testWorld.TestSubaccountAddr = addr
	return nil
}

func (s *AccountSteps) requestAccountOverview() error {
	subAddr := s.testWorld.TestSubaccountAddr
	if subAddr == "" {
		subAddr = "0xtest_subaccount_address"
	}

	client, err := s.testWorld.GetReadClient()
	if err != nil {
		return err
	}

	overview, err := client.GetAccountOverview(subAddr)
	if err != nil {
		s.testWorld.SetError(err)
		return err
	}

	s.testWorld.AccountOverview = overview
	return nil
}

func (s *AccountSteps) shouldReceiveAccountOverview() error {
	if s.testWorld.LastError != nil {
		return fmt.Errorf("expected no error, got: %v", s.testWorld.LastError)
	}
	if s.testWorld.AccountOverview == nil {
		return fmt.Errorf("account overview should be set")
	}
	return nil
}

func (s *AccountSteps) checkTotalMargin() error {
	if s.testWorld.LastError != nil {
		return nil
	}
	if s.testWorld.AccountOverview.TotalMargin < 0 {
		return fmt.Errorf("total margin should be non-negative")
	}
	return nil
}

func (s *AccountSteps) checkUnrealizedPnl() error {
	if s.testWorld.LastError != nil {
		return nil
	}
	// Unrealized PnL can be negative, so just check it's finite
	_ = s.testWorld.AccountOverview.UnrealizedPnl
	return nil
}

func (s *AccountSteps) checkCrossMarginRatio() error {
	if s.testWorld.LastError != nil {
		return nil
	}
	if s.testWorld.AccountOverview.CrossMarginRatio < 0 {
		return fmt.Errorf("cross margin ratio should be non-negative")
	}
	return nil
}

func (s *AccountSteps) requestPositions() error {
	subAddr := s.testWorld.TestSubaccountAddr
	if subAddr == "" {
		subAddr = "0xtest_subaccount_address"
	}

	client, err := s.testWorld.GetReadClient()
	if err != nil {
		return err
	}

	positions, err := client.GetUserPositions(subAddr)
	if err != nil {
		s.testWorld.SetError(err)
		return err
	}

	s.testWorld.Positions = positions
	return nil
}

func (s *AccountSteps) shouldReceivePositions() error {
	if s.testWorld.LastError != nil {
		return fmt.Errorf("expected no error, got: %v", s.testWorld.LastError)
	}
	// positions could be empty if the account has no open positions
	return nil
}

func (s *AccountSteps) requestOpenOrders() error {
	subAddr := s.testWorld.TestSubaccountAddr
	if subAddr == "" {
		subAddr = "0xtest_subaccount_address"
	}

	client, err := s.testWorld.GetReadClient()
	if err != nil {
		return err
	}

	orders, err := client.GetUserOpenOrders(subAddr)
	if err != nil {
		s.testWorld.SetError(err)
		return err
	}

	s.testWorld.OpenOrders = orders
	return nil
}

func (s *AccountSteps) shouldReceiveOpenOrders() error {
	if s.testWorld.LastError != nil {
		return fmt.Errorf("expected no error, got: %v", s.testWorld.LastError)
	}
	// open_orders could be empty
	return nil
}

File: sdk-rust/tests/bdd/steps/account_steps.rs
//! Step definitions for account-management.feature

use cucumber::{then, when};
use decibel_sdk::error::DecibelError;

use crate::TestWorld;

#[when("I request the account overview for a subaccount")]
async fn request_account_overview(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_account_overview(&sub_addr, None, None).await {
        Ok(overview) => world.account_overview = Some(overview),
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the account overview data")]
async fn should_receive_account_overview(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
    assert!(world.account_overview.is_some(), "Account overview should be set");
}

#[then("the overview should include the total margin")]
async fn check_total_margin(world: &mut TestWorld) {
    assert!(!world.has_error());
    let _overview = world.account_overview.as_ref().expect("Account overview should be set");
}

#[then("the overview should include the unrealized PnL")]
async fn check_unrealized_pnl(world: &mut TestWorld) {
    assert!(!world.has_error());
    let _overview = world.account_overview.as_ref().expect("Account overview should be set");
}

#[then("the overview should include the cross margin ratio")]
async fn check_cross_margin_ratio(world: &mut TestWorld) {
    assert!(!world.has_error());
    let _overview = world.account_overview.as_ref().expect("Account overview should be set");
}

#[when(expr = "I request all subaccounts for account {string}")]
async fn request_subaccounts(world: &mut TestWorld, account_addr: String) {
    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_subaccounts(&account_addr).await {
        Ok(_subaccounts) => {},
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive a list of subaccounts")]
async fn should_receive_subaccounts(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[then("each subaccount should have a subaccount address")]
async fn check_subaccount_address(world: &mut TestWorld) {
    assert!(!world.has_error());
}

#[when("I request the account positions")]
async fn request_positions(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_positions(&sub_addr, None, None, None).await {
        Ok(positions) => world.positions = Some(positions),
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the account positions")]
async fn should_receive_positions(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[when("I request the account open orders")]
async fn request_open_orders(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_open_orders(&sub_addr).await {
        Ok(orders) => world.open_orders = Some(orders),
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the open orders")]
async fn should_receive_open_orders(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[when("I request the account trade history")]
async fn request_trade_history(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_trade_history(&sub_addr, None, None).await {
        Ok(_trades) => {},
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the trade history")]
async fn should_receive_trade_history(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[when("I request the account funding history")]
async fn request_funding_history(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_funding_history(&sub_addr, None, None, None).await {
        Ok(_history) => {},
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the funding history")]
async fn should_receive_funding_history(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[when("I request the account deposit and withdrawal history")]
async fn request_fund_history(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_user_fund_history(&sub_addr, None, None).await {
        Ok(_history) => {},
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the deposit and withdrawal history")]
async fn should_receive_fund_history(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}

#[when("I request the delegations for a subaccount")]
async fn request_delegations(world: &mut TestWorld) {
    let sub_addr = world.test_subaccount_addr.clone()
        .unwrap_or_else(|| "0xtest_subaccount_address".to_string());

    if world.get_or_create_read_client().is_err() {
        return;
    }
    let client = world.read_client.as_ref().unwrap();
    match client.get_delegations(&sub_addr).await {
        Ok(_delegations) => {},
        Err(e) => world.set_error(e),
    }
}

#[then("I should receive the delegation list")]
async fn should_receive_delegations(world: &mut TestWorld) {
    assert!(!world.has_error(), "Expected no error, got: {:?}", world.last_error);
}
